nilearn/glm/first_level/first_level.py

```diff
@@ -979,108 +971,656 @@ def first_level_from_bids(dataset_path, task_label, space_label=None,
+
+    Parameters
+    ----------
+    confounds : :obj:`list` of :obj:`str`
+        List of fullpath to the confounds.tsv files
+
+    imgs : :obj:`list` of :obj:`str`
+        List of fullpath to the preprocessed images
+
+    """
+    if confounds and len(confounds) != len(imgs):
+        raise ValueError(
+            f"{len(confounds)} confounds.tsv files found "
+            f"for {len(imgs)} bold files. "
+            "Same number of confound files as "
+            "the number of runs is expected"
+        )
+
+
+def _validate_args_first_level_from_bids(
+    dataset_path: str | Path,
+    task_label: str,
+    space_label: str | None,
+    sub_labels: list[str] | None,
+    img_filters: list[tuple[str, str]],
+    derivatives_folder: str,
+) -> None:
+    """Check type and value of arguments of first_level_from_bids.
+
+    Check that:
+        - dataset_path is a string and exists
+        - derivatives_path exists
+        - task_label and space_label are valid bids labels
+        - img_filters is a list of tuples of strings
+          and all filters are valid bids entities
+          with valid bids labels
+
+    Parameters
+    ----------
+    dataset_path : :obj:`str`
+        Fullpath of the BIDS dataset root folder.
+
+    task_label : :obj:`str`
+        Task_label as specified in the file names like _task-<task_label>_.
+
+    space_label : :obj:`str`
+        Specifies the space label of the preprocessed bold.nii images.
+        As they are specified in the file names like _space-<space_label>_.
+
+    sub_labels : :obj:`list` of :obj:`str`, optional
+        Specifies the subset of subject labels to model.
+        If 'None', will model all subjects in the dataset.
+
+    img_filters : :obj:`list` of :obj:`tuples` (str, str)
+        Filters are of the form (field, label).
+        Only one filter per field allowed.
+
+    derivatives_path : :obj:`str`
+        Fullpath of the BIDS dataset derivative folder.
+
+    """
+    if not isinstance(dataset_path, (str, Path)):
+        raise TypeError(
+            "'dataset_path' must be a string or pathlike. "
+            f"Got {type(dataset_path)} instead."
+        )
+    dataset_path = Path(dataset_path)
+    if not dataset_path.exists():
+        raise ValueError(f"'dataset_path' does not exist:\n{dataset_path}")
+
+    if not isinstance(derivatives_folder, str):
+        raise TypeError(
+            "'derivatives_folder' must be a string or pathlike. "
+            f"Got {type(derivatives_folder)} instead."
+        )
+    derivatives_folder = dataset_path / derivatives_folder
+    if not derivatives_folder.exists():
+        raise ValueError(
+            "derivatives folder not found in given dataset:\n"
+            f"{derivatives_folder}"
+        )
+
+    _check_bids_label(task_label)
+
+    if space_label is not None:
+        _check_bids_label(space_label)
+
+    if not isinstance(sub_labels, list):
+        raise TypeError(
```

Comment:
and here

Reply:
Done.

I noticed there are a couple of other places that are not covered but several of them belong to a part that is refactored in #3605 so I would rather not had them in this PR.